lib/drivers/oneTouchUltra2.js

```diff
@@ -0,0 +1,386 @@
+      listenForPacket(18000, function(err, pkt) {
+        if (err === 'TIMEOUT') {
+        } else {
+          callback(err, pkt);
+        }
+      });
+
+    });
+
+  };
+
+  var getSomeInfo = function (obj, cb) {
+      console.log('DEBUG: on getSomeInfo');
+      cb();
+  };
+
+  var prepBGData = function (progress, data) {
+    cfg.builder.setDefaults({ deviceId: data.id });
+    var dataToPost = [];
+
+    //This is converting to a string the data bytes array
+    var lines = String.fromCharCode.apply(null, data.bytes);
+
+    //Separe each data row by \n
+    var splites = lines.split('\n');
+
+    // extract header
+    var hpattern = /P (\d{3}),"(\w{9})","(\w+\/\w+).*" (\w+)/;
+
+    var toMatch = hpattern.exec(splites[0]);
+    var nrecords = toMatch[1],
+        serialno = toMatch[2],
+        oum = toMatch[3],
+        checksum = toMatch[4],
+        calcchks = 0;
+
+    //set missing serial number and id
+    data.serialNumber = serialno;
+    data.id = data.id + serialno;
+
+    // var bytes = data.bytes.subarray(0, 27, 0);
+    var stringWithoutChk = 'P ' + nrecords + ',"' + serialno + '","' + oum + ' "';
+    var bytes = [0];
+
+    for (var i = 0; i < stringWithoutChk.length; ++i)
+    {
+        bytes.push(stringWithoutChk.charCodeAt(i));
+    }
+
+    calcchks = _.reduce(bytes, function(s1, e){ return s1 + e;}, 0);
+
+    if(calcchks !== parseInt(checksum, 16)){
+         console.log('hmmm...bad checksum', calcchks, checksum);
+         return null;
+     }
+
+    // get data from lines
+    var dpattern = /P "(\w+)","(\d{2}\/\d{2}\/\d{2})","(\d{2}:\d{2}:\d{2})   ","..(\d{3}).","(.)","(\d{2})",\W00\W(.+)/;
+    _.forEach(splites.slice(1), function(l){ // <-- jump the header
+        var ms = l.match(dpattern);
+        if(ms){
+            var dow = ms[1], // day of week
+                dor = ms[2], // date of reading
+                tor = ms[3], // time of reading
+                rf  = ms[4], // result
+                alv = ms[5], // alpha value
+                umc = ms[6], // numeric value for user meal comment
+                chk = ms[7]; // checksum
+                createchk = 0;
+
+            var lineWithoutChecksum = 'P "' + dow + '","'+ dor + '","' + tor + '   ", "' + rf + ' ","' + alv + '","' + umc + '", 00 ';
+            var arrayBytes = [0];
+
+            for (var i = 0; i < lineWithoutChecksum.length; ++i)
+            {
+                arrayBytes.push(lineWithoutChecksum.charCodeAt(i));
+            }
+
+            createchk = _.reduce(arrayBytes, function(s1, e){ return s1 + e;}, 0);
+
+            if(createchk !== parseInt(chk, 16)){
+              console.log('hmmm...bad checksum', createchk, chk);
+              return null;
+            } else {
+              var bdt = buildDateTime(dor, tor);
+
+              var smbg = cfg.builder.makeSMBG()
+                              .with_deviceId('OneTouchUltra2 ' + serialno)
+                              //.with_deviceSerialNumber(serialno)
```

Comment:
That's why there are separate steps for fetchData, processData, and uploadData -- you should be doing the assembly of data in the processData step, so that you have all the information you need at that point.